// react-use-toggle-hook/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useToggle = void 0;
var react_1 = require("react");
var react_use_when_mounted_1 = require("react-use-when-mounted");
var useToggle = function (initialValue) {
    if (initialValue === void 0) { initialValue = false; }
    var whenMounted = react_use_when_mounted_1.useWhenMounted();
    var _a = react_1.useState(initialValue), value = _a[0], setValue = _a[1];
    var open = react_1.useCallback(function () { return whenMounted(function () { return setValue(true); }); }, [setValue, whenMounted]);
    var close = react_1.useCallback(function () { return whenMounted(function () { return setValue(false); }); }, [setValue, whenMounted]);
    var toggle = react_1.useCallback(function () { return whenMounted(function () { return setValue(function (p) { return !p; }); }); }, [setValue, whenMounted]);
    return {
        value: value,
        toggle: toggle,
        open: open,
        close: close,
        on: open,
        off: close,
    };
};
exports.useToggle = useToggle;
